### DISCUSSION TITLE
Push github stored docker container image to Google Container Repository

### DISCUSSION THREAD

**Discussion by dschinkel** [community/community#26030] [Actions]
<p>I’ve generated a docker package to the GitHub Container repository for my project.  Now I want to take and push it up to Google Container Repository.</p>
<p>I’m referencing <a href="https://github.com/GoogleCloudPlatform/github-actions/blob/df7d4c84a1b7d23e96e46eb6922b0462cd77df70/example-workflows/cloud-run/.github/workflows/cloud-run.yml" rel="nofollow noopener">this example</a> for starters.</p>
<p><strong>I tweaked it a bit and tried this:</strong></p>
<pre><code class="lang-auto">  push_docker_image_to_gcloud_container_registry:
    needs: setup_gcloud_cli
    runs-on: ubuntu-latest
    name: Push Docker image to Google Container Registry
    steps:
      - run: gcloud submit --quiet --tag "gcr.io/$PROJECT_ID/$SERVICE_NAME:$GITHUB_SHA"
</code></pre>
<p><strong>I took the <em>build</em> part out</strong> of  the script this thinking I could just <strong>submit</strong> my <strong>image</strong> since I already built my image with github actions prior to this.  But this doesn’t work.</p>
<p>How do I just tell this to push my already existing package/image to GC?</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/e98921fda4eef6e523b522ba53fd457cfade7b1a.jpeg" data-download-href="/uploads/short-url/xjX3u9ZjRZFr1yf6284Fe5pYbN0.jpeg?dl=1" title="Screen Shot 2020-09-01 at 12.10.53 AM copy" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/338798/181094865-85680eea-06c3-4c61-a52b-808fe6e61de2.jpeg" alt="Screen Shot 2020-09-01 at 12.10.53 AM copy" data-base62-sha1="xjX3u9ZjRZFr1yf6284Fe5pYbN0" width="690" height="238" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/e/e98921fda4eef6e523b522ba53fd457cfade7b1a_2_690x238.jpeg, https://aws1.discourse-cdn.com/github/original/2X/e/e98921fda4eef6e523b522ba53fd457cfade7b1a.jpeg 1.5x, https://aws1.discourse-cdn.com/github/original/2X/e/e98921fda4eef6e523b522ba53fd457cfade7b1a.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/e98921fda4eef6e523b522ba53fd457cfade7b1a_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-09-01 at 12.10.53 AM copy</span><span class="informations">1000×345 120 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p><img src="https://user-images.githubusercontent.com/338798/181094866-864363ef-5bec-4397-8f90-9d4f422ee42e.jpeg" alt="Screen Shot 2020-09-01 at 12.13.33 AM" data-base62-sha1="aQIAHiapfN7tzH8GGTUVInHr8j9" width="600" height="225"></p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/dschinkel">@dschinkel</a> ,</p>
<p>You can try with the following steps:</p>
<ol>
<li>
<p>Login to GitHub Packages (<strong>docker login</strong>), and pull (<strong>docker pull</strong>) the image from your GitHub repository.</p>
<pre><code class="lang-auto">docker login docker.pkg.github.com -u USERNAME -p PASSWORD
docker pull docker.pkg.github.com/OWNER/REPOSITORY/IMAGE_NAME:TAG
</code></pre>
</li>
<li>
<p>Get the <strong>IMAGE_ID</strong> via the “<a href="https://docs.docker.com/engine/reference/commandline/images/" rel="nofollow noopener"><strong>docker images</strong></a>” command, and set new image name and tag (<strong>docker tag</strong>) with your Google Cloud Container Registry URL for the image.</p>
<pre><code class="lang-auto">image_id=$(docker images docker.pkg.github.com/OWNER/REPOSITORY/IMAGE_NAME:TAG | awk 'NR == 2 {print $3}')
docker tag $image_id gcr.io/PROJECT-ID/IMAGE_NAME:TAG
</code></pre>
</li>
<li>
<p>Login to the Google Cloud Container Registry (if need), and push (<strong>docker push</strong>) the image to your Google Cloud Container Registry.</p>
</li>
</ol>

**Comment by dschinkel** [comment]
<p>Alright you nudged me to the right direction at least, thanks!</p>
<p>But I do want to use the new <a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/migrating-to-github-container-registry-for-docker-images" rel="nofollow noopener">GitHub Docker Container Repository</a>.</p>
<h1>Old Way: GitHub Packages Docker registry</h1>
<h3><a href="https://docs.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-docker-for-use-with-github-packages#installing-a-package" rel="nofollow noopener">Installing a package</a></h3>
<ul>
<li>This is for the older <strong>GitHub Docker packages</strong> which I think is the example you were showing @ <a href="https://github.community/u/brightran">brightran</a>
</li>
</ul>
<blockquote>
<p>You can use the  <code>docker pull</code>  command to install a docker image from GitHub Packages, replacing  <em>OWNER</em>  with the name of the user or organization account that owns the repository,  <em>REPOSITORY</em>  with the name of the repository containing your project,  <em>IMAGE_NAME</em>  with name of the package or image,  <em>TAG_NAME</em>  with tag for the image you want to install</p>
</blockquote>
<blockquote>
<p><strong>Note:</strong>  You must pull the image using  <code>IMAGE_NAME:VERSION</code>  and not using  <code>IMAGE_NAME:SHA</code></p>
</blockquote>
<h3><a href="https://docs.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-docker-for-use-with-github-packages#authenticating-with-the-github_token" rel="nofollow noopener">Authenticating with the  <code>GITHUB_TOKEN</code> </a></h3>
<blockquote>
<p>If you are using a GitHub Actions workflow, you can use a  <code>GITHUB_TOKEN</code>  to publish and consume packages in GitHub Packages without needing to store and manage a personal access token. For more information, see "[Authenticating with the  <code>GITHUB_TOKEN</code> ].</p>
</blockquote>
<h3><a href="https://docs.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token#about-the-github_token-secret" rel="nofollow noopener">About the  <code>GITHUB_TOKEN</code>  secret</a></h3>
<blockquote>
<p>GitHub automatically creates a  <code>GITHUB_TOKEN</code>  secret to use in your workflow. You can use the  <code>GITHUB_TOKEN</code>  to authenticate in a workflow run.</p>
</blockquote>
<blockquote>
<p>When you enable GitHub Actions, GitHub installs a GitHub App on your repository. The  <code>GITHUB_TOKEN</code>  secret is a GitHub App installation access token</p>
</blockquote>
<h1>New Way: GitHub Container Registry</h1>
<h3><a href="https://docs.github.com/en/packages/getting-started-with-github-container-registry/migrating-to-github-container-registry-for-docker-images#authenticating-with-the-container-registry" rel="nofollow noopener">Authenticating with the container registry</a></h3>
<p>You will need to authenticate to the container registry with the base URL  <code>ghcr.io</code> . We recommend creating a new access token for using the container registry.</p>
<p>Using the CLI for your container type, sign in to the GitHub Container Registry service at  <code>ghcr.io</code> .</p>
<pre><code class="lang-auto">$ echo $CR_PAT | docker login ghcr.io -u USERNAME --password-stdin
</code></pre>
<h3><a href="https://docs.github.com/en/packages/managing-container-images-with-github-container-registry/pushing-and-pulling-docker-images" rel="nofollow noopener">Pushing and pulling Docker images</a></h3>
<p>Well this sucks, I have to use a personal access token.  You can’t just use <code>GITHUB_TOKEN</code> to auth like you could with the old Docker Packages:</p>
<blockquote>
<p>GitHub Container Registry does not support using  <code>GITHUB_TOKEN</code>  for your PAT so you must use a different custom variable, such as  <code>CR_PAT</code></p>
</blockquote>
<p>So something like this which would have worked in the old flow, won’t work with the new:</p>
<pre><code class="lang-auto">$ echo $CR_PAT | docker login ghcr.io -u ${{ secrets.GITHUB_TOKEN }} --password-stdin
</code></pre>
<p>You have to <a href="https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token" rel="nofollow noopener">create a token</a>, store it, then call:</p>
<pre><code class="lang-auto">echo $CR_PAT | docker login ghcr.io -u USERNAME --password-stdin
</code></pre>
<p>I’ll Circle back once I try some things with the new Container Registry and see if I can get it working.</p>
